xarray/core/utils.py

```diff
@@ -66,6 +66,14 @@ def _maybe_cast_to_cftimeindex(index: pd.Index) -> pd.Index:
         return index
 
 
+def maybe_cast_to_coords_dtype(label, coords_dtype):
+    if isinstance(label, float):
+        label = coords_dtype.type(label)
+    elif isinstance(label, list) and coords_dtype.kind == 'f':
```

Comment:
We also want to convert arrays like `np.array([0, 1/3, 2/3, 1], dtype=np.float32)`.

Actually, maybe we could actually convert any indexing argument to a float array of the appropriate dtype? e.g.,
```python
def maybe_cast_to_coords_dtype(label, coords_dtype):
    if coords_dtype.kind == 'f':
        label = np.asarray(label, dtype=coords_dtype)
    return label
```

Reply:
This is definitely cleaner. Thank you. However, we still need to check label is not of type slice.

Reply:
I adopted the change in the latest code, but added a check for the case of label is slice